docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,494 @@
+- However, allow a control plane using an external etcd cluster to scale up to other numbers such as 2 or 4.
+- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec, should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
+- Scale up operations should not be done in conjunction with an upgrade operation.
+- Scale up operations *could* be blocked based on the Health status
+  - for stacked etcd quorum must be considered
+  - for external etcd only availability of etcd should be considered.
+
+![controlplane-init-6](images/controlplane/controlplane-init-6.png)
+
+##### Scale Down
+
+- Scale down a stacked control plane
+  - For each replica being deleted:
+    - Remove the replica’s etcd member from the etcd cluster
+    - Update the kubeadm generated config map
+- Scale down a control plane with an external etcd
+  - Same as above minus Etcd management
+- Scale down operations should not be done in conjunction with an upgrade operation, this should not impact manual operations for recovery.
+- Scale down operations *could* be blocked based on the Health status
+  - for stacked etcd quorum must be considered
+  - for external etcd only availability of etcd should be considered.
+- Scale to 0 should be rejected for the initial support
+
+![controlplane-init-7](images/controlplane/controlplane-init-7.png)
+
+##### Delete of the entire ControlPlane (kubectl delete controlplane my-controlplane)
+
+- Completely removing the control plane and issuing a delete on the underlying machines.
+  - Deletion is equivalent to a scale to 1 followed by a deletion of a single replica Control Plane.
+
+##### Cluster upgrade (using create-swap-and-delete)
+
+- Triggered on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used.
+  - Must be able to trigger an upgrade for the underlying machine image (e.g. nothing in the above list has changed but an upgrade is required)
+    - Possible implementations include (not necessary to hammer out exact implementation right now):
+      - Manually bumping a Generation field on the ControlPlane object
+      - Adding a field that is not the Generation field for users to manually bump to trigger an upgrade
+      - Adding a timestamp field, something like `nextUpgrade` which can be set to a specific time in the future (far future if no upgrades are desired) and an upgrade will run when that timestamp is passed. Good for scheduling upgrades/SLOs & quickly modifying to NOW or yesterday if an upgrade is required immediately. Would potentially update the field value to some interval (2 months from now?)
+- An upgrade will look like this:
+  - Serially go through each control plane replica not at the current config hash
+    - Check health of any replicas for the current config hash
```

Comment:
Same as above, there is no field or indication how to get or expose health information for a control plane